WikipediaUnitTests/Code/LocationManagerTests.swift

```diff
@@ -0,0 +1,230 @@
+    func testExistingHeadingWithoutCallback() {
+        let heading = MockCLHeading(headingAccuracy: 10)
+        mockCLLocationManager.simulateUpdate(heading: heading)
+        XCTAssertEqual(locationManager.heading, heading) // Test last know heading on location manager
+    }
+
+    func testStopUpdating() {
+        locationManager.startMonitoringLocation()
+        // Setup value when monitoring
+        let location1 = CLLocation(latitude: 10, longitude: 20)
+        mockCLLocationManager.simulateUpdate(location: location1)
+        let heading1 = MockCLHeading(headingAccuracy: 10)
+        mockCLLocationManager.simulateUpdate(heading: heading1)
+
+        locationManager.stopMonitoringLocation()
+        // Update with new values after stop monitoring for location
+        let location2 = CLLocation(latitude: 100, longitude: 200)
+        mockCLLocationManager.simulateUpdate(location: location2)
+        let heading2 = MockCLHeading(headingAccuracy: 100)
+        mockCLLocationManager.simulateUpdate(heading: heading2)
+
+        // Heading and location
+        XCTAssertEqual(locationManager.heading, heading1)
+        XCTAssertEqual(delegate.heading, heading1)
+        XCTAssertEqual(locationManager.location, location1)
+        XCTAssertEqual(delegate.location, location1)
+
+        // Error
+        let error = NSError(domain: "org.wikimedia.wikipedia.test", code: -1, userInfo: nil)
+        mockCLLocationManager.simulate(error: error)
+        XCTAssertNil(delegate.error)
+    }
+
+    func testReceiveError() {
+        locationManager.startMonitoringLocation()
+
+        let error = NSError(domain: "org.wikimedia.wikipedia.test", code: -1, userInfo: nil)
+        mockCLLocationManager.simulate(error: error)
+        XCTAssertEqual((delegate.error as NSError?), error)
+    }
+
+    func testChangeAuthorizedState() {
+        mockCLLocationManager.update(authorizationStatus: .authorizedAlways)
+        mockCLLocationManager.delegate?.locationManager?(CLLocationManager(), didChangeAuthorization: .denied)
+        XCTAssertEqual(delegate.authorized, false)
+
+        mockCLLocationManager.update(authorizationStatus: .denied)
+        mockCLLocationManager.delegate?.locationManager?(CLLocationManager(), didChangeAuthorization: .authorizedAlways)
+        XCTAssertEqual(delegate.authorized, true)
+    }
+
+    // MARK: - Test heading
+
+    func testUpdateDeviceHeading() {
+        locationManager.startMonitoringLocation()
+        let heading = MockCLHeading(headingAccuracy: 10)
+        mockCLLocationManager.simulateUpdate(heading: heading)
+        mockDevice.simulateUpdate(orientation: .unknown)
+
+        mockDevice.simulateUpdate(orientation: .portrait)
+        XCTAssertEqual(mockCLLocationManager.headingOrientation, .portrait)
+
+        mockDevice.simulateUpdate(orientation: .landscapeLeft)
+        XCTAssertEqual(mockCLLocationManager.headingOrientation, .landscapeLeft)
+
+        locationManager.stopMonitoringLocation()
```

Comment:
// Device orientation updates should not be propagated when the monitoring is stopped.